Downloads/Uploads UI: Fix "Last activity" column not persisting restarts, add "Last failure"

xor-freenet opened the pull request:
(Please use the following as merge commit message)

# Motivation
As documented at https://bugs.freenetproject.org/view.php?id=6384 (which you don't have to read since this message contains all important information from it), previously, the "Last activity" column of the "Downloads" / "Uploads" section of the web interface did not persist restarts. This was because:
- class ClientRequest, which tracked the time of the last activity, updated the activity when it received a SplitFileProgressEvent.
- SplitFileProgressEvents are *not* only sent when a real progress happens. To not have to write initialization code, fred generates fake events at startup to initialize stuff.

This had to be fixed because it is critical for usability to tell when downloads have no progress anymore: Files fall out of the Freenet network frequently, and to be able to purge them from your download queue, you need to be able to see "this hasn't had any progress for weeks".

# Features
This branch aims to fix the "Last activity" column to persist restarts. It does so by:
- Moving tracking of the "Last activity" Date from class ClientRequest to class ClientRequester.  The class where it is moved to is what generates the SplitfileProgressEvents, so it can properly track the date independent of those events, and even actually propagate it inside them.
- Changing SplitFileProgressEvent to include the actual "Last activity" date to transport it to all higher levels of code
- Exposing the date in class RequestStatus, which is what the UI actually takes its data from.

# Bonus features
- A "Last failure" column is added to the UI as well. This column tells the last time when the variables ClientRequester.failedBlocks or ClientRequester.fatallyFailedBlocks last changed. It is only visible in advanced mode of the web interface. Suggested usage of it is to help people to debug bugs of fred which cause stalling downloads (or even uploads): As long as fred is working properly, either "Last activity" or "Last progress" should continuously be updated. If both fail to change, then fred is failing to request data from the network, which is a stalling bug.
- As a consequence of adding "Last failure", the "Last activity" column is renamed to "Last progress" to distinguish the two columns in a better way.
- English and German l10n is included.
- The last progress / last failure dates are also exposed in the FCP message "SimpleProgress".

# Compatibility
This branch had previously caused loss of uploads for people who had used purge-db4o testing releases such as testing-build-1468-pre1 already. This should no longer be the case. All downloads and uploads should be preserved.
People who have already used a version of this branch before https://github.com/xor-freenet/fred-staging/commit/8c43801c3039c86321c21a20da68f91d1516bddb will lose their uploads though. (There is no easy way for me to fix this: Fred doesn't keep a non-serialization backup of uploads, so there is nothing I could restore them from.)

# Testing
There are 3 types of transfers which you need to test:
- Downloads
- Uploads of single files
- Uploads of whole directories

Also, test both the simple and the advanced mode of the web interface.

# Known Bugs
- The "Last failure" column always showed "never" in all my tests, even for downloads which I would have expected to have suffered DataNotFound.  I think this is not a bug in the code which this branch adds, but rather in preexisting fred code. Fred probably somehow fails to call ClientRequester.failedBlock() when there is a DataNotFound error.  I have asked toad to review this branch for whether I am right with that, and for ideas how to fix it.

Thynix commented:
I appreciate the effort behind such an in-depth merge commit message / pull request description, but it includes so much information I have trouble determining what parts of it are important for me to read before I start review. This seems to include in-depth thoughts about things that came up during development, and ideas about future directions. This is great for a development log or a bug tracker entry but I don't think this much material belongs here.

It might also be good to have a guideline for how much can fit in a commit message. Those are pretty long too. How about this - if it exceeds 200 words or so, it might be including too much detail and/or not be written concisely.

Thoughts?

xor-freenet commented:
> I appreciate the effort behind such an in-depth merge commit message / pull request description, but it includes so much information I have trouble determining what parts of it are important for me to read before I start review.

You can skip reading the sections "Testing" and "Known bugs" if you want to. 
You can also skip reading the linked bugtracker entry 6384 which induced this pull request - the pull request message aims to contain all its information so people don't have to read that.
I've just amended the part which links the bugtracker entry with "which you don't have to read since this message contains all important information from it" so this is apparent. Sorry.

The length is also because the class hierarchy of the stuff which this modifies is rather complex: I just counted, there are 28 classes to represent requests in that area of code. It took me several hours of reading it before I was able to write a single line of code :| Thus I thought it's necessary to explain an overview of which classes the bug resulted from (at Motivation), and how the fix works (at Features).

But to honor your request to shorten it, I have just reviewed the PR message, and was able to remove 1 sentence. No ideas what else to remove, sorry.
If you can teach me about something which is too much text, feel free to do so. It is always easier to respond to specific criticism than a generic "too long, didn't read" :)

> It might also be good to have a guideline for how much can fit in a commit message. Those are pretty long too.

Please provide feedback on which ones are too long and why.
I ask for that because when squashing, I did already review each commit message to remove duplication which resulted from squashing similar commits, the messages are not merely concatenations. 
The squashing was from 13 to 6 commits, which with over 50% is a lot, so thats a possible explanation for the length. I guess since this is my first squashing session, I might have overachieved to your disappointment. If you give me examples on what you would have split up into multiple commits, I can try to learn from that for the future :)

>  How about this - if it exceeds 200 words or so, it might be including too much detail and/or not be written concisely.

Where does this magic number come from? Is there for example a document which analyzes the average commit message size of large, successful software projects?

ArneBab commented:
As I understand it, the comment from @Thynix was that this is too long to review. A development log is nice, but it shouldn’t be required to read it to be able to review the code.

> I have trouble determining what parts of it are important for me to read before I start review.

@xor-freenet I think @Thynix needs an answer to this.

xor-freenet commented:
> As I understand it, the comment from @Thynix was that this is too long to review. A development > log is nice, but it shouldn’t be required to read it to be able to review the code.
>
>>    I have trouble determining what parts of it are important for me to read before I start review.
>
>@xor-freenet I think @Thynix needs an answer to this.

I think you missed my comment right before your? Summary: The length has been reduced in honor of @Thynix complaint already.

The pull request main message is not a one-liner since it also serves the purpose of providing a pre-written message for putting in the merge commit, to summarize the work of the branch so reviewers can review the branch as a whole diff instead of reviewing individual commits.

xor-freenet commented:
> As I understand it, the comment from @Thynix was that this is too long to review. A development > log is nice, but it shouldn’t be required to read it to be able to review the code.
>
>>    I have trouble determining what parts of it are important for me to read before I start review.
>
>@xor-freenet I think @Thynix needs an answer to this.

I think you missed my comment right before yours? Summary: The length has been reduced in honor of @Thynix complaint already. The comment also clarifies which parts do not have to be read before review.

(The pull request main message is not a one-liner since it also serves the purpose of providing a pre-written message for putting in the merge commit, to summarize the work of the branch so reviewers can review the branch as a whole diff instead of reviewing individual commits.)

ArneBab commented:
I saw your comment, but even the shortened version is really long. Imagine coming from work and deciding to have a take at reviewing some code before dinner. Or taking 15 minutes out of the lunch break for that.

impression when I looked into this pull-request was “oof, that’s a lot of text”.

There’s almost one line of explanation in the commit-messages for every two lines of code. For very complex code this can be the right thing to do. But here it looks redundant: Do people need this much text to judge whether your code does what it should?

In my case, I expect an at most two sentence description of the changes, so I can quickly check whether the code fits that.

Thynix commented:
I'm suggesting shorter commit bodies not to suggest that these commit messages change necessarily - more to demonstrate for future commits what I consider to be enough information.

Thynix commented on src/freenet/clients/fcp/RequestStatus.java:
```diff
@@ -63,10 +69,15 @@ synchronized void restart(boolean started) {
 		this.minBlocks = min;
 		this.fetchedBlocks = fetched;
+        // clone() because Date is mutable
+		this.latestSuccess
```
The whitespace here is inconsistent - some lines are indented with tabs and others with spaces. You are free to keep it as-is if you like.

xor-freenet replied:
I am aware of this, it is due to Eclipse bugs.

I had tried to manually fix the resulting inconsistencies during the work on plugin-fcp-rewrite, but this showed that it usually takes more time than writing the actual code. This is because Eclipse sometimes automatically converts tabs/spaces when pasting or pressing "tab", so editing the wrong whitespace out is not easy.

I can upgrade my Eclipse to see whether this is fixed, but I would be glad if I could postpone this until my bachelor thesis is over (3 months). I don't think it would be a good idea to modify my working environment while the thesis is running.

Merging such code meanwhile should not wreak much havoc: We used to recommend people to configure their IDE to 4-spaces-per-tab, and with that this won't be visible. It is only visible here because Github uses something else than 4-spaces-per-tab.

-> ACK to ignore such stuff for 3 months?

Thynix replied:
That one's editor has bugs hardly seems like a sufficient excuse to write messy whitespace, and one cannot rule out that people set their editors to display whitespace explicitly. (Spoilers: I do!) If you don't want to fix this I can do it after we merge this.

xor-freenet replied:
>  If you don't want to fix this I can do it after we merge this.

Sorry if I didn't make this clear enough: The problem is not my willingness. I *am* willing to fix such things.
The problem is that I have doubt whether I can justify causing this to cost FPI money by doing it myself: I have tried doing this during plugin-fcp-rewrite, and the amount of time it consumed was almost as much as writing the code took.

You're part of the people who audit the bills, so I would like to ask you to decide now whether I should do this or not. I'll follow your decision :)

For your decision making, more information:

> and one cannot rule out that people set their editors to display whitespace explicitly. (Spoilers: I do!)

Java is not white-space sensitive such as Python. Do we need to support it if people wrongly configure their editors to behave as if it was?

Thynix replied:
Whitespace can fail to meet the standards even if people do not configure their editors to display it.

xor-freenet replied:
Yes well thats OK. The most important question for me is whether I can justify billing time to fix such stuff. Whats your stance on that?

xor-freenet replied:
Whitespace fixed by 3418c7daa580672fa0239b41113ac27ec388f2d2.
Done this since I assume that people prefer having 1468 soon to having an answer about whether billing probably less than 5 EUR here is OK.

Thynix commented on src/freenet/clients/fcp/RequestStatus.java:
```diff
@@ -136,17 +153,15 @@ public boolean isStarted() {
+		this.latestFailure = event.latestFailure != null ? (Date)event.latestFailure.clone() : null;
 		this.fetchedBlocks = event.succeedBlocks;
+        // clone() because Date is mutable
+		this.latestSuccess = event.latestSuccess != null ? (Date)event.latestSuccess.clone() : null;
```
Whitespace is inconsistent here too.

xor-freenet replied:
Whitespace fixed by 3418c7daa580672fa0239b41113ac27ec388f2d2.
Done this since I assume that people prefer having 1468 soon to having an answer about whether billing probably less than 5 EUR here is OK.

Thynix commented on src/freenet/clients/fcp/SimpleProgressMessage.java:
```diff
@@ -35,7 +35,9 @@ public SimpleFieldSet getFieldSet() {
 		fs.put("Required", event.minSuccessfulBlocks);
 		fs.put("Failed", event.failedBlocks);
 		fs.put("FatallyFailed", event.fatallyFailedBlocks);
+		fs.put("LatestFailure", event.latestFailure != null ? event.latestFailure.getTime() : 0);
```
Inconsistent whitespace here.

Thynix replied:
Whitespace is inconsistent here - it's added with spaces below and tabs here.

xor-freenet replied:
Fixed by 624ab1610da28c36572b82ed8614512692607b37

Thynix commented on src/freenet/clients/fcp/SimpleProgressMessage.java:
```diff
@@ -70,6 +72,11 @@ public double getFetchedBlocks(){
 	
+	public Date getLatestSuccess() {
+        // clone() because Date is mutable
+	    return event.latestSuccess != null ? (Date)event.latestSuccess.clone() : null;
```
Inconsistent whitespace here.

xor-freenet replied:
Whitespace fixed by 3418c7daa580672fa0239b41113ac27ec388f2d2.
Done this since I assume that people prefer having 1468 soon to having an answer about whether billing probably less than 5 EUR here is OK.

Thynix commented on src/freenet/clients/fcp/RequestStatus.java:
```diff
@@ -63,10 +69,15 @@ synchronized void restart(boolean started) {
 		this.totalBlocks = total;
 		this.minBlocks = min;
 		this.fetchedBlocks = fetched;
+        // clone() because Date is mutable
```
What's your reasoning for including this comment? It's certainly the right behavior, but I don't think it's remarkable enough to warrant a comment.

ArneBab replied:
I expect that this should reduce the maintenance burden in case people who don’t know much Java try to improve the codebase. It wouldn’t have been obvious to me that `clone()` is needed.

xor-freenet replied:
What @ArneBab said :)
-> ACK to keep as is?

Thynix replied:
Fine.

Thynix commented on src/freenet/clients/fcp/ClientGet.java:
```diff
@@ -925,14 +924,17 @@ public boolean filterData() {
 	synchronized RequestStatus getStatus() {
 		boolean totalFinalized = false;
 		int total = 0, min = 0, fetched = 0, fatal = 0, failed = 0;
+		Date latestSuccess = CurrentTimeUTC.get(), latestFailure = null;
```
Multiple variables in a single declaration is [not allowed by the coding standards](https://google-styleguide.googlecode.com/svn/trunk/javaguide.html#s4.8.2.1-variables-per-declaration).

xor-freenet replied:
> Why does latestSuccess start at the current time? Is my assumption correct that getStatus() may be called (and indicate progress) without progressPending being set?

This is explained elsewhere:
```
     * Initialized to current time for usability purposes: This allows the user to sort downloads by
     * last success in the user interface to determine which ones are stalling - those will be the
     * ones with the oldest last success date. If we initialized it to "null" only, that would not
     * be possible: The user couldn't distinguish very old stalling downloads from downloads which
     * merely had no success yet because they were added a short time ago */
```

I will add that comment to all the places where it is actually initialized.


> Multiple variables in a single declaration is not allowed by the coding standards.

Will fix. This was done to be coherent with the previous line.

xor-freenet replied:
> Multiple variables in a single declaration is not allowed by the coding standards.

Fixed by 3317ff517ce124ac8dbb9c14ed02ae5fb625a9fc.

xor-freenet replied:
> Why does latestSuccess start at the current time? Is my assumption correct that getStatus() may be called (and indicate progress) without progressPending being set?

Commit 2cea4dd24242562d1e3b54951a8b0ac462f26dee fixes this and similar occurrences to point the reader to the JavaDoc which explains the reason for the default value.

Thynix commented on src/freenet/clients/fcp/ClientPut.java:
```diff
@@ -519,6 +521,7 @@ RequestStatus getStatus() {
 		if(fnam != null) fnam = new File(fnam.getPath());
 		
 		int total=0, min=0, fetched=0, fatal=0, failed=0;
+		Date latestSuccess = CurrentTimeUTC.get(), latestFailure = null;
```
Multiple declarations in one line here too.

xor-freenet replied:
Fixed by 3317ff517ce124ac8dbb9c14ed02ae5fb625a9fc

Thynix commented on src/freenet/clients/fcp/ClientPutDir.java:
```diff
@@ -376,6 +378,7 @@ RequestStatus getStatus() {
 		}
 		
 		int total=0, min=0, fetched=0, fatal=0, failed=0;
+		Date latestSuccess = CurrentTimeUTC.get(), latestFailure = null;
```
Multiple declarations here too.

xor-freenet replied:
Fixed by 3317ff517ce124ac8dbb9c14ed02ae5fb625a9fc

Thynix commented on src/freenet/clients/fcp/ClientRequest.java:
```diff
@@ -27,6 +27,9 @@
 
+    /**
+     * ATTENTION: When incrementing this, please skip version 2. Version 2 had already temporarily
+     * been used by a development branch. */
```
What's the reasoning for putting effort into supporting code that was never released?

xor-freenet replied:
I had initially implemented this like you demand here: The initial version bumped the UID to 2, so the serialization code would not eat the files; and then fallback to the backup which is stored in a different format. (Nice that toad implemented this!)

This would have however caused loss of uploads since the backup code is only implemented for downloads.
I did not care about that since we have no reason to support compatibility for testing releases as you said.

But @ArneBab rightfully argued that purge-db4o beta versions have been out for a very long time and many people might be using them.

I would suggest the following solution:
- I refrain from casting a vote here since I already had a heavy discussion with @ArneBab  and I think he won. You instead discuss with @ArneBab  whether to keep as is or drop compatibility.
- To avoid such issues in the future, make a decision whether testing releases may causs loss of downloads/etc. Also, always warn people about potential loss of downloads/uploads with testing releases in the release notes.

Thynix replied:
It's true that the purge-db4o snapshots / preleases have been out a long time - my question is why blacklist UID 2 when its usage was never even merged?

xor-freenet replied:
UID 2 had been in use by the initial version of this pull request, the commit to remove it was only added after some time.
I had announced the pull request on IRC, and thus other people might have been using it already.
So I'd rather be save than sorry. There are more than enough version UIDs, its a long :)

Thynix replied:
I'm not concerned about running out of `long`s. I don't think it's reasonable to bother to document compatibility with highly unstable, unreleased code like this. If someone was running unreleased code and want to avoid it they can edit the code themselves - adding this comment increases clutter and the effort required to change the UID. (Is there even a situation where we're willing to change the UID? I thought that broke backwards compatibility?)

If you still disagree I will give up on this because I want to get this over with.

xor-freenet replied:
OK, I've removed the excessive documentation at f1f71fefbd78fb25bb1ee556ae5528e1fe16d217

> (Is there even a situation where we're willing to change the UID? I thought that broke backwards compatibility?)

The purpose of UIDs is to break compatibility, yes. This is intended, and scheduled for regular use: If we ever need to change the class architecture in a way which would be incompatible with pre-existing serialized data, then we would change the UID. This would ensure that the deserializer would refuse to load the existing serialized data. The node would then falll back to the backup of the download queue which it stores in a different format. 
The reason why the UID changing was not used in this case is that the backup code has the disadvantage of only being implemented for the download-queue, not for the upload queue.
I would by the way suggest that we soon implement backup for uploads as well to prevent future huge discussions of whether to change or not change the UID. We've already had this discussion now even though the code was only released as testing version. In the future, it could very well be necessary that we break compatibility to the released official code.

Thynix replied:
I don't object to backups or upgrades for more things.

xor-freenet replied:
OK nice, I therefore filed a bug to implement upload queue backup: https://bugs.freenetproject.org/view.php?id=6533

Thynix commented on src/freenet/clients/fcp/ClientPut.java:
```diff
@@ -527,17 +530,19 @@ RequestStatus getStatus() {
-				succeeded, total, min, fetched, fatal, failed, totalFinalized, 
-				lastActivity, priorityClass, finalURI, uri, failureCode,
-				failureReasonShort, failureReasonLong, getDataSize(), mimeType,
-				fnam, isCompressing());
```
May I register my displeasure that these arguments are all broken apart? This is a huge list, and this change perpetuates it. (Not that its intent was to clean it up.)

xor-freenet replied:
I assume you meant to comment this at the whole changed constructor function signature, not just at the single point where it is called here?
In case that is true, here's my reply:

The intentions of changing the signature like that were:
- Add the new parameter at a reasonable position in the parameter list, not just to the end. The "latestSuccess" is added after the "fetched" parameter, which is the number of successful blocks. The "latestFailure" is added after the "failed" parameter.
- To obey the line length limit, the parameters are shifted around to the next lines.

I assume what you complain about is the shifting of all parameters to obey the line length?
Do you see any other possibility of introducing new parameters in the middle without shifting while also obeying the line length limit?
I could of course insert a new line which contains nothing but the new parameter, but wouldn't that be ugly to the point where the current solution is preferable?

Thynix replied:
This is more a complaint about the large argument list in the existing codebase than something to change here.

Thynix commented on src/freenet/clients/fcp/RequestStatus.java:
```diff
@@ -102,9 +113,15 @@ public int getFetchedBlocks() {
 		return fetchedBlocks;
 	}
 
-	public long getLastActivity() {
```
This is an API change. Please add backwards compatibility.

xor-freenet replied:
Fixed by bebb3ca5bcc877df1e07a0f754447d30bfde7779

Thynix commented on src/freenet/clients/fcp/RequestStatus.java:
```diff
@@ -102,9 +113,15 @@ public int getFetchedBlocks() {
 
-	public long getLastActivity() {
-		return lastActivity;
+	public Date getLastActivity() {
```
Thoughts on naming this `getLastSuccess()` or `getLastProgress()` instead?

xor-freenet replied:
Fixed by bebb3ca5bcc877df1e07a0f754447d30bfde7779

Thynix commented on src/freenet/l10n/freenet.l10n.en.properties:
```diff
@@ -1616,9 +1616,12 @@ QueueToadlet.insertFileBrowseLabel=Upload a file from disk
 QueueToadlet.insertFileResetForm=Reset Form
 QueueToadlet.key=Key
-QueueToadlet.lastActivity=Last Activity
+QueueToadlet.lastActivity=Last Progress
```
Thoughts on changing the key to `QueueToadlet.lastProgress`?

xor-freenet replied:
First notice what I answered to the below question of yours:

> > The success date is listed as "progress." Why not have the variable names be lastProgress to match? (FCP too.)
>
> This is because I unfortunately realized that "progress" is better *after* I wrote the code :( 
> It would be quite a bit of work to refactor the code since "success" is used in many places, so I thought I could use the common practice of renaming stuff by l10n only in case where the code is too much.
> 
> Do you want me to fix this or keep as is?

My answer depends what your answer is there.
If  we keep using "activity" for variables in the code, while having "Progress" in the l10n string, then I would say that we should keep using "activity" for the l10n key as well: The workaround of having a mismatch between l10n value and in-code variable naming should be apparent from the l10n file, since the l10n file is the place where it was started. Mismatching the key name and value there is the only way to indicate this mismatch, since comments are not possible in l10n files (or are they?)
Do you agree with that?

Thynix replied:
Okay, that makes sense. Comments are possible in some l10n formats but not this one.

Thynix commented on src/freenet/clients/http/QueueToadlet.java:
```diff
@@ -44,10 +45,11 @@
 import freenet.client.filter.ContentFilter;
-import freenet.client.filter.KnownUnsafeContentTypeException;
 import freenet.client.filter.FilterMIMEType;
+import freenet.client.filter.KnownUnsafeContentTypeException;
```
This should not have included reorganizing imports - it contains no formatting commits.

xor-freenet replied:
The Date import is new, which is why I organized the imports to get it. This can be done in one step with Eclipse: After you typed code which uses a class which is not imported yet, you press CTRL+O. It then imports the new class, and also organizes the remaining imports.
I usually don't even notice doing this anymore.

I'd say the time which is saved by this is worth the occasional noise. We also get a gradual, implicit cleanup of imports by doing this frequently, similar to the gradual fixing of programming standards by applying them to new code. 
-> ACK to keep as is?

Thynix replied:
I'm okay with keeping this currently but I'd rather that future commits - where practical - not mix functional and formatting changes.

Thynix commented on src/freenet/clients/http/QueueToadlet.java:
```diff
@@ -1353,7 +1355,17 @@ else if(secondFinalized && !firstFinalized)
-						result = Fields.compare(firstRequest.getLastActivity(), secondRequest.getLastActivity());
+                        Date a = firstRequest.getLastActivity();
+                        Date b = secondRequest.getLastActivity();
+                        if(a == null) a = new Date(0);
```
The [coding standards require](https://google-styleguide.googlecode.com/svn/trunk/javaguide.html#s4.6.2-horizontal-whitespace) a space between the `if` and `(`.

xor-freenet replied:
> This code is duplicated. Thoughts on moving it out to a function which allows null?
> EDIT: What about adding Date support to Fields.compare()?

Fixed by 88c73464bd82234b05ca6113481d9ad9131c4ed3

> The coding standards require a space between the if and (. They also do not allow the omission of braces here.

Fixed by 9e0d129a4c2eaafe3872a63e1da0f154667b8574

> Why Date(0) on null?

Added cb4b91dc904fa36e61510676f5aa5fedb3f7bec5 to explain this.
Also notice that the Date objects can be null here because if the user was on a purge-db4o testing release before his serialized data would not contain the Date fields. That part was not added to the comment in the above commit since the Fields class is unrelated to the "Last activity" stuff and thus shouldn't document its quirks.

Thynix commented on src/freenet/clients/http/QueueToadlet.java:
```diff
@@ -1652,7 +1664,17 @@ else if(secondFinalized && !firstFinalized)
+                        new QueueColumn[] {
+                            QueueColumn.IDENTIFIER, QueueColumn.PRIORITY, QueueColumn.SIZE,
+                            QueueColumn.MIME_TYPE, QueueColumn.PROGRESS, QueueColumn.LAST_ACTIVITY,
+                            QueueColumn.LAST_FAILURE, QueueColumn.PERSISTENCE, QueueColumn.FILENAME,
```
It's hard to tell what changed here when functional changes are combined with formatting ones.

Thynix commented on src/freenet/clients/http/QueueToadlet.java:
```diff
@@ -1978,15 +2019,32 @@ private void selectLocation(HTMLNode node) {
-		if (lastActivity == 0) {
+		if (lastActivity == null) {
+            // This is "unknown"  instead of "never" because the backend of RequestStatus always
+            // tries to initialize lastActivity to current time, it should probably never be null.
```
That sounds like it makes that an error, not just "unknown." Is that worthy of logging?

xor-freenet replied:
Fixed by 5eaf2c8e2bacc7277e273919d30890839a272084

Thynix commented:
The success date is listed as "progress." Why not have the variable names be `lastProgress` to match?

Thynix commented on src/freenet/clients/fcp/ClientRequest.java:
```diff
@@ -295,16 +295,6 @@ public boolean isStarted() {
-	 *
-	 * @return The time of the request’s last activity, or {@code 0}
-	 */
-	public long getLastActivity() {
```
Removing this breaks API.

xor-freenet replied:
Fixed by b1dbc1f5a04460e386ccfa635b69f48d9693bbb5

Thynix commented:
The proposed merge commit message could probably be reduced to

   This adds persistent tracking of last progress and last failure to
    requests, and exposes it in the queue pages and FCP SimpleProgress
    messages.

Thoughts?

xor-freenet commented:
> The success date is listed as "progress." Why not have the variable names be lastProgress to match? (FCP too.)

This is because I unfortunately realized that "progress" is better *after* I wrote the code :( 
It would be quite a bit of work to refactor the code since "success" is used in many places, so I thought I could use the common practice of renaming stuff by l10n only in case where the code is too much.

Do you want me to fix this or keep as is?

Thynix commented:
That makes sense. I'm fine with just changing the l10n.

xor-freenet commented:
Progress summary:
- All requested code changes are fixed or pending a reply to questions I've asked.
- I've written up a reply which explains why I would keep the main merge commit message as I suggested it. However, due to the huge annoyance I've caused during previous code review, I will give myself 1-3 days of time to think about whether what I want to say is acceptable before I actually post it :)
- Until the discussion about the code changes is finished, I'll not deal with rewriting history / re-squashing to fix the individual commit messages as suggested: The code changes might cause more stuff which should be squashed into existing commits, so it makes no sense to rewrite history before.

xor-freenet commented:
> I've written up a reply which explains why I would keep the main merge commit message as I suggested it. However, due to the huge annoyance I've caused during previous code review, I will give myself 1-3 days of time to think about whether what I want to say is acceptable before I actually post it :)

..but here's one good news as a sneak peek: I will definitely refrain from casting a vote/veto, you guys of course decide what's in the merge commit message; all I'll say will be just for your information to help in the decision making procedure :)

Thynix commented:
I forgot about the FCP fields. I'm on board with keeping the variables and l10n as-is, but the FCP fields are entirely new. Thoughts on naming it LastProgress?

xor-freenet commented:
> I forgot about the FCP fields. I'm on board with keeping the variables and l10n as-is, but the FCP fields are entirely new. Thoughts on naming it LastProgress there?

Fixed by 2681f6999770d1c5ac554e3c782dad97f278d351

xor-freenet commented:
> The proposed merge commit message could probably be reduced to

> > This adds persistent tracking of last progress and last failure to
> > requests, and exposes it in the queue pages and FCP SimpleProgress
> > messages.

> EDIT: Maybe along with explanation of what events are considered failures.

> Thoughts?


DISCLAIMER: What follows is NOT to be considered as a vote/veto, I refrain from casting one. Due to the recent annoyance I have caused during fred review, I would rather just provide my opinion as "for your information" only, and let you guys decide on your own what you do.
I will try to keep future merge commit message suggestions even shorter. The fact that the first request to shorten it actually caused me to find stuff to remove shows that I am still bad at writing short messages and need to practice more. I will try to do that :)
______

As far as I've understood the various contributions to the code review debate on devl, there are 2 good options:
1. Review branches as a whole diff, not as individual commits. This requires a more detailled merge commit message since the reviewers will not read the commit messages of the individual commits and thus don't have their information.
2. Review branches as individual commits. This requires squashing commits so there is less of them. It also requires more detailled commit messages since they are the primary entry point, not the branch merge commit message.


So if one of those strategys is used, either the individual commit messages or the whole branch merge commit message could be considered to be simplified a lot since only the opposite will be read.

Now it seems to me you requested both shortifying the merge commit message *and* the individual commit messages.
This will leave both types of review at a poor level of information.

So IMHO, you should do one of those:
- decide for one review strategy, and thus refrain from requesting simplification of the messages needed for the other. I don't favor this option since the following option seems better to me
- realize that there might be future additional reviews of the code, for which you cannot control the used strategy since you will not be doing them but someone else. Thus, decide that it would be good to allow *both* strategies by allowing both extensive individual commit messages and a long merge commit message. Then you should also stick to only reading one of the two, so you don't get overwhelmed by the sum of their verboseness. You may then of course also refuse to merge pull requests if one of the two message types is too short.


Also, additional information which doesn't quite fit in the train of though I just mentioned but is important nevertheless:
The intent behind the merge commit messages was to be fully self-contained. This means that it should make sense to future readers even if:
- Github is bankrupt, and thus the original pull request message is gone.
- We one day decide to use a different bugtracker software and only migrate non-resolved issues. (I am strongly against doing any of this.)
- The future readers have not participated in the huge discussions which explained the motivation behind the pull request
- The future readers have not read the > 28 classes which are involved in the code which tracks downloads/uploads.

This intent has already yielded some result: While processing the review results, for 2 times I was able to look up information in the merge commit message which I had forgotten after some weeks already and needed for fixing the review results.


So overall, you maybe might want to try to keep as much of the current information in the merge commit message as possible.
But again, please be assured that I am perfectly fine with whatever solution you think is the best; and that I will try to learn to improve the shortness of my messages.

xor-freenet commented:
New progress summary:
- All requested code changes are fixed or pending a reply to questions I've asked.
- I will wait for a final request to rewrite the commits before I do that, to give time for the review of the pending changes to be finished.

Thynix commented:
Given that Last Failure isn't in the GUI I'd be up for keeping it out of FCP too.

Thynix commented:
Are you open to an appeal to authority? The Linux kernel is a huge project which uses git. Its commit messages look like [this](https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=7a6cb0abe1aa63334f3ded6d2b6c8eca80e72302) and its merge commit messages look like [this](https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=0a68c6bc7ce9d4855f3130f9eff8ff774b597531). I feel like in contrast the detail that's currently here is overwhelming.

Thynix commented:
If you're okay with it I'd like to stop with this back-and-forth and merge something based on this.

ArneBab commented:
I’d suggest to just write short messages and put the explanations into a blog. If you like writing that much explanation. When I look up a commit, I’m not searching for prose. I’m searching for short snippets of information with what I need.

But since the first line of the commits is OK (needed for low-overhead creation of release notes), I would just merge the commits as they are and request shorter messages in future. I will not read commit messages if they are this long.

To understand this, please imagine coming back home, getting off the bike and taking 15 minutes to look at the most recent pull-request. You know that you’ll need to start making dinner at 19:00, but you still have 15 minutes. So you look into the pull-request and the explanations are longer than the code changes. You can either read half the explanations or just look at the code. So you try to understand what the code does and what it intends to do from the code alone. After 15 minutes you post a partial review and start cooking. Next slot for code review is tomorrow evening, or maybe next friday. The pull-request lies open for several weeks while more changes pile up.

Contrast that with short commit messages: You look into the pull-request. The commit message gives you the intention of the change (“sounds good”), maybe with a short note on non-obvious side-effects of the implementation, and you skim the code to see whether it realizes the intention. If it does and you don’t see problems which the writer might have overlooked: Great, code review finished. You write the review and go make dinner. The pull-request is merged the same week.

xor-freenet commented:
> Given that Last Failure isn't in the GUI I'd be up for keeping it out of FCP too.

Fixed by 624ab1610da28c36572b82ed8614512692607b37.
Bugtracker entry also updated to mention this: https://bugs.freenetproject.org/view.php?id=6526

xor-freenet commented:
> If you're okay with it I'd like to stop with this back-and-forth and merge something based on this.

Yes I'd also like to get this over with!
For your convenience, I've fixed everything remaining *including* the whitespace issues.
Thus, I'd say feel free to modify commit structure/messages as you please; but please notify me about additional code changes so I can review them, as all code change requests should be done already.

Thanks for your effort in getting this done :)

Thynix commented:
Alright, cool! I expect to have more time after work today.

On Tue, Jun 9, 2015, 9:27 AM xor-freenet <notifications@github.com> wrote:

> If you're okay with it I'd like to stop with this back-and-forth and merge
> something based on this.
>
> Yes I'd also like to get this over with!
> For your convenience, I've fixed everything remaining *including* the
> whitespace issues.
> Thus, I'd say feel free to modify commit structure/messages as you please;
> but please notify me about additional code changes so I can review them, as
> all code change requests should be done already.
>
> Thanks for your effort in getting this done :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/freenet/fred/pull/341#issuecomment-110356616>.
>

Thynix commented:
Merged in adca3fb. Thanks!
